File: Windows/src/Services/DashScopeClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;
using ClipForgeAI.Win.Models;

namespace ClipForgeAI.Win.Services;

public sealed class ApiException : Exception
{
    public int StatusCode { get; }
    public ApiException(int status, string message) : base(message) { StatusCode = status; }
}

/// <summary>
/// DashScope REST 客户端(声音克隆 / OSS 上传 / 视频任务提交 + 轮询)。
/// 端点、文案与 macOS 客户端保持完全一致,便于跨平台行为统一。
/// </summary>
public sealed class DashScopeClient
{
    private static readonly HttpClient Http = new() { Timeout = TimeSpan.FromMinutes(10) };
    private static readonly JsonSerializerOptions JsonOpts = new() { PropertyNamingPolicy = null };

    private readonly string _apiKey;
    public DashScopeClient(string apiKey) { _apiKey = apiKey; }

    private HttpRequestMessage Req(HttpMethod method, string url, bool async = false,
        IDictionary<string, string>? extra = null, HttpContent? body = null)
    {
        var r = new HttpRequestMessage(method, url);
        r.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _apiKey);
        r.Headers.Add("user-agent", "clipforge");
        if (async) r.Headers.Add("X-DashScope-Async", "enable");
        if (extra != null)
            foreach (var kv in extra) r.Headers.TryAddWithoutValidation(kv.Key, kv.Value);
        if (body != null) r.Content = body;
        return r;
    }

    private static async Task<JsonElement> SendAsync(HttpRequestMessage req, CancellationToken ct = default)
    {
        var resp = await Http.SendAsync(req, ct);
        var body = await resp.Content.ReadAsStringAsync(ct);
        if (!resp.IsSuccessStatusCode)
            throw new ApiException((int)resp.StatusCode, $"HTTP {(int)resp.StatusCode}: {body}");
        return string.IsNullOrWhiteSpace(body) ? default : JsonDocument.Parse(body).RootElement.Clone();
    }

    // ============= 声音克隆(create_voice + 轮询) =============

    public async Task<string> CreateVoiceAsync(string prefix, string audioUrl, CancellationToken ct = default)
    {
        var body = new
        {
            model = FixedModel.VoiceEnrollment,
            input = new
            {
                action = "create_voice",
                target_model = FixedModel.Tts,
                prefix,
                url = audioUrl
            }
        };
        using var req = Req(HttpMethod.Post,
            $"{FixedModel.Endpoint}/services/audio/tts/customization",
            body: JsonContent(body));
        var json = await SendAsync(req, ct);
        return json.GetProperty("output").GetProperty("voice_id").GetString()!;
    }

    public async Task<JsonElement> QueryVoiceAsync(string voiceId, CancellationToken ct = default)
    {
        using var req = Req(HttpMethod.Get,
            $"{FixedModel.Endpoint}/services/audio/tts/customization?voice_id={Uri.EscapeDataString(voiceId)}");
        return await SendAsync(req, ct);
    }

    public async Task<JsonElement> ListVoicesAsync(string prefix, CancellationToken ct = default)
    {
        using var req = Req(HttpMethod.Get,
            $"{FixedModel.Endpoint}/services/audio/tts/customization?prefix={Uri.EscapeDataString(prefix)}");
        return await SendAsync(req, ct);
    }

    // ============= OSS 临时上传(用于本地文件给 DashScope 引用) =============

    public async Task<string> UploadToOssAsync(string localPath, CancellationToken ct = default)
    {
        // 1) 申请策略
        var policyReq = Req(HttpMethod.Post, $"{FixedModel.Endpoint}/uploads");
        policyReq.Content = JsonContent(new
        {
            model = FixedModel.Tts,
            input = new { action = "get_policy" }
        });
        var policy = await SendAsync(policyReq, ct);
        var upload = policy.GetProperty("output").GetProperty("upload");
        var host = upload.GetProperty("host").GetString()!;
        var policyB64 = upload.GetProperty("policy").GetString()!;
        var accessKeyId = upload.GetProperty("access_key_id").GetString()!;
        var accessKeySecret = upload.GetProperty("access_key_secret").GetString()!;
        var securityToken = upload.GetProperty("security_token").GetString()!;
        var ossKey = upload.GetProperty("oss_key").GetString()!;
        var expiration = upload.GetProperty("expiration").GetString()!;

        // 2) 走 OSS POST(multipart/form-data)
        using var form = new MultipartFormDataContent();
        form.Add(new StringContent(accessKeyId), "OSSAccessKeyId");
        form.Add(new StringContent(policyB64), "policy");
        form.Add(new StringContent("200"), "SignatureVersion");
        form.Add(new StringContent("OSS-HMAC-SHA256"), "SignatureMethod");
        form.Add(new StringContent("oss"), "dir");
        form.Add(new StringContent(expiration), "expire");
        form.Add(new StringContent(ossKey), "key");
        form.Add(new StringContent(securityToken), "x-oss-security-token");
        form.Add(new StringContent("True"), "success_action_status");

        var fileBytes = await File.ReadAllBytesAsync(localPath, ct);
        var fileContent = new ByteArrayContent(fileBytes);
        fileContent.Headers.ContentType = new MediaTypeHeaderValue("application/octet-stream");
        form.Add(fileContent, "file", Path.GetFileName(localPath));

        var uploadResp = await Http.PostAsync(host, form, ct);
        var uploadBody = await uploadResp.Content.ReadAsStringAsync(ct);
        if (!uploadResp.IsSuccessStatusCode)
            throw new ApiException((int)uploadResp.StatusCode, $"OSS 上传失败: {uploadBody}");

        // 3) 通知 DashScope 资源就绪
        var submitReq = Req(HttpMethod.Post, $"{FixedModel.Endpoint}/uploads");
        submitReq.Content = JsonContent(new
        {
            model = FixedModel.Tts,
            input = new { action = "submit", oss_key = ossKey }
        });
        var submitJson = await SendAsync(submitReq, ct);
        var resourceUrl = submitJson.GetProperty("output").GetProperty("resource").GetString()!;
        return resourceUrl;
    }

    // ============= 视频生成(异步任务) =============

    public async Task<string> SubmitVideoGenAsync(VideoGenRequest req, CancellationToken ct = default)
    {
        var input = new Dictionary<string, object>
        {
            ["prompt"] = req.Prompt,
        };
        if (!string.IsNullOrEmpty(req.ReferenceImagePath)) input["img_url"] = req.ReferenceImagePath;
        if (!string.IsNullOrEmpty(req.ReferenceAudioPath)) input["audio_url"] = req.ReferenceAudioPath;

        var parameters = new Dictionary<string, object>
        {
            ["resolution"] = req.Resolution,
            ["prompt_extend"] = req.PromptExtend,
            ["duration"] = req.Duration,
            ["shot_type"] = req.ShotType,
        };

        var body = new { model = FixedModel.VideoI2V, input, parameters };
        using var msg = Req(HttpMethod.Post,
            $"{FixedModel.Endpoint}/services/aigc/video-generation/video-synthesis",
            async: true,
            body: JsonContent(body));
        var json = await SendAsync(msg, ct);
        return json.GetProperty("output").GetProperty("task_id").GetString()!;
    }

    public async Task<JsonElement> GetTaskAsync(string taskId, CancellationToken ct = default)
    {
        using var req = Req(HttpMethod.Get, $"{FixedModel.Endpoint}/tasks/{Uri.EscapeDataString(taskId)}");
        return await SendAsync(req, ct);
    }

    public async Task DownloadAsync(string url, string destPath, CancellationToken ct = default)
    {
        using var resp = await Http.GetAsync(url, HttpCompletionOption.ResponseHeadersRead, ct);
        resp.EnsureSuccessStatusCode();
        await using var stream = await resp.Content.ReadAsStreamAsync(ct);
        Directory.CreateDirectory(Path.GetDirectoryName(destPath)!);
        await using var fs = File.Create(destPath);
        await stream.CopyToAsync(fs, ct);
    }

    private static StringContent JsonContent(object o)
    {
        var s = JsonSerializer.Serialize(o, JsonOpts);
        return new StringContent(s, Encoding.UTF8, "application/json");
    }
}

File: Sources/DashScopeClient.swift
import Foundation

struct APIError: LocalizedError {
    let message: String
    init(_ m: String) { message = m }
    var errorDescription: String? { message }
}

enum JSONValue {
    static func string(_ obj: Any?, _ keyPath: [String]) -> String? {
        var cur = obj
        for k in keyPath {
            guard let d = cur as? [String: Any], let v = d[k] else { return nil }
            cur = v
        }
        if let s = cur as? String { return s }
        if let n = cur as? NSNumber { return n.stringValue }
        return nil
    }
    static func dict(_ obj: Any?, _ key: String) -> [String: Any]? {
        (obj as? [String: Any])?[key] as? [String: Any]
    }
}

@MainActor
final class DashScopeClient {
    static let shared = DashScopeClient()
    private let session: URLSession
    private init() {
        let cfg = URLSessionConfiguration.default
        cfg.timeoutIntervalForRequest = 60
        cfg.timeoutIntervalForResource = 3600
        session = URLSession(configuration: cfg)
    }

    private func key() throws -> String {
        let k = AppSettings.shared.apiKey.trimmingCharacters(in: .whitespaces)
        if k.isEmpty { throw APIError("请先在「设置」中填写 DashScope API Key") }
        return k
    }

    private func authHeaders(extra: [String: String] = [:]) throws -> [String: String] {
        var h = ["Authorization": "Bearer \(try key())"]
        for (k, v) in extra { h[k] = v }
        return h
    }

    // MARK: - OSS 临时上传（供 img_url / audio_url 使用）

    /// 上传本地文件到 DashScope 临时 OSS，返回 oss:// 形式的 URL
    func uploadToOSS(model: String, fileURL: URL) async throws -> String {
        // 1) getPolicy
        var comp = URLComponents(string: "\(DashScope.httpBase)/uploads")
        comp?.queryItems = [URLQueryItem(name: "action", value: "getPolicy"),
                            URLQueryItem(name: "model", value: model)]
        guard let policyURL = comp?.url else { throw APIError("uploads URL 无效") }
        var req = URLRequest(url: policyURL)
        req.httpMethod = "GET"
        for (k, v) in try authHeaders() { req.setValue(v, forHTTPHeaderField: k) }
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        let (pdata, presp) = try await session.data(for: req)
        guard let http = presp as? HTTPURLResponse, (200..<300).contains(http.statusCode) else {
            throw APIError("获取上传凭证失败：\(String(data: pdata, encoding: .utf8) ?? "HTTP \( (presp as? HTTPURLResponse)?.statusCode ?? -1 )")")
        }
        let policyJson = try JSONSerialization.jsonObject(with: pdata) as? [String: Any] ?? [:]
        guard let output = policyJson["data"] as? [String: Any] ?? policyJson["output"] as? [String: Any] else {
            throw APIError("上传凭证响应缺少 data 字段")
        }
        guard let host = output["upload_host"] as? String,
              let dir = output["upload_dir"] as? String,
              let policy = output["policy"] as? String,
              let sig = output["signature"] as? String,
              let ak = output["oss_access_key_id"] as? String else {
            throw APIError("上传凭证字段不完整")
        }

        // 2) POST 到 OSS
        let fileName = fileURL.lastPathComponent
        let key = dir + "/" + fileName
        guard let ossURL = URL(string: host) else { throw APIError("upload_host 无效") }
        var boundary = "----ClipForge\(UUID().uuidString)"
        let mime = mimeFor(fileURL)

        var body = Data()
        func field(_ name: String, _ value: String) {
            body.append(Data("--\(boundary)\r\n".utf8))
            body.append(Data("Content-Disposition: form-data; name=\"\(name)\"\r\n\r\n".utf8))
            body.append(Data("\(value)\r\n".utf8))
        }
        field("OSSAccessKeyId", ak)
        field("Signature", sig)
        field("policy", policy)
        field("key", key)
        field("x-oss-object-acl", (output["x_oss_object_acl"] as? String) ?? "private")
        field("x-oss-forbid-overwrite", (output["x_oss_forbid_overwrite"] as? String) ?? "true")
        field("success_action_status", "200")
        field("x-oss-content-type", mime)
        // file field last
        let fileData = try Data(contentsOf: fileURL)
        body.append(Data("--\(boundary)\r\n".utf8))
        body.append(Data("Content-Disposition: form-data; name=\"file\"; filename=\"\(fileName)\"\r\n".utf8))
        body.append(Data("Content-Type: \(mime)\r\n\r\n".utf8))
        body.append(fileData)
        body.append(Data("\r\n--\(boundary)--\r\n".utf8))
        _ = boundary

        var oreq = URLRequest(url: ossURL)
        oreq.httpMethod = "POST"
        oreq.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        oreq.setValue("application/json", forHTTPHeaderField: "Accept")
        let (oBody, oResp) = try await session.upload(for: oreq, from: body)
        guard let ohttp = oResp as? HTTPURLResponse, (200..<300).contains(ohttp.statusCode) else {
            throw APIError("上传 OSS 失败：\(String(data: oBody, encoding: .utf8) ?? "HTTP \( (oResp as? HTTPURLResponse)?.statusCode ?? -1 )")")
        }
        return "oss://" + key
    }

    private func mimeFor(_ url: URL) -> String {
        switch url.pathExtension.lowercased() {
        case "png": return "image/png"
        case "jpg", "jpeg": return "image/jpeg"
        case "webp": return "image/webp"
        case "wav": return "audio/wav"
        case "mp3": return "audio/mpeg"
        case "m4a": return "audio/mp4"
        case "mp4": return "video/mp4"
        default: return "application/octet-stream"
        }
    }

    // MARK: - 声音复刻 enrollment

    func createVoice(targetModel: String, prefix: String, url: String) async throws -> String {
        let body: [String: Any] = [
            "model": FixedModel.voiceEnrollment,
            "input": ["action": "create_voice", "target_model": targetModel,
                      "prefix": prefix, "url": url],
        ]
        var extra: [String: String] = [:]
        if url.hasPrefix("oss://") { extra["X-DashScope-OssResourceResolve"] = "enable" }
        let json = try await postJSON("\(DashScope.httpBase)/services/audio/tts/customization",
                                      body: body, extra: extra)
        guard let vid = JSONValue.string(json["output"], ["voice_id"]) else {
            throw APIError("创建音色失败：\(json["code"] ?? "") \(json["message"] ?? "")")
        }
        return vid
    }

    func queryVoice(voiceId: String) async throws -> [String: Any] {
        let body: [String: Any] = [
            "model": FixedModel.voiceEnrollment,
            "input": ["action": "query_voice", "voice_id": voiceId],
        ]
        let json = try await postJSON("\(DashScope.httpBase)/services/audio/tts/customization", body: body)
        return json["output"] as? [String: Any] ?? [:]
    }

    func listVoices(prefix: String) async throws -> [[String: Any]] {
        let body: [String: Any] = [
            "model": FixedModel.voiceEnrollment,
            "input": ["action": "list_voice", "prefix": prefix, "page_index": 0, "page_size": 50],
        ]
        let json = try await postJSON("\(DashScope.httpBase)/services/audio/tts/customization", body: body)
        return (json["output"] as? [String: Any])?["voice_list"] as? [[String: Any]] ?? []
    }

    func deleteVoice(voiceId: String) async throws {
        let body: [String: Any] = [
            "model": FixedModel.voiceEnrollment,
            "input": ["action": "delete_voice", "voice_id": voiceId],
        ]
        _ = try await postJSON("\(DashScope.httpBase)/services/audio/tts/customization", body: body)
    }

    // MARK: - 视频生成任务 (wan2.6-i2v)

    struct VideoRequest {
        var model: String = FixedModel.videoI2V
        var prompt: String
        var imageURL: String?          // 文生视频(t2v)不需要首帧图
        var audioURL: String?
        var resolution: String = "720P"
        var duration: Int = 10
        var promptExtend: Bool = true
        var audioEnabled: Bool = true
        var shotType: String = "multi"
    }

    func submitVideoTask(_ r: VideoRequest) async throws -> String {
        var input: [String: Any] = ["prompt": r.prompt]
        if let img = r.imageURL, !img.isEmpty { input["img_url"] = img }
        if let a = r.audioURL { input["audio_url"] = a }
        var params: [String: Any] = [
            "resolution": r.resolution,
            "prompt_extend": r.promptExtend,
            "duration": r.duration,
            "audio": r.audioEnabled,
        ]
        if r.shotType == "multi" { params["shot_type"] = "multi" }
        let body: [String: Any] = ["model": r.model, "input": input, "parameters": params]
        let json = try await postJSON(
            "\(DashScope.httpBase)/services/aigc/video-generation/video-synthesis",
            body: body,
            extra: ["X-DashScope-Async": "enable", "X-DashScope-OssResourceResolve": "enable"])
        guard let tid = JSONValue.string(json["output"], ["task_id"]) else {
            throw APIError("提交视频任务失败：\(json["code"] ?? "") \(json["message"] ?? "")")
        }
        return tid
    }

    struct VideoStatus {
        var state: String
        var videoUrl: String?
        var message: String?
        var raw: [String: Any]
    }

    func pollVideoTask(_ taskId: String) async throws -> VideoStatus {
        let url = URL(string: "\(DashScope.httpBase)/tasks/\(taskId)")!
        var req = URLRequest(url: url)
        req.httpMethod = "GET"
        for (k, v) in try authHeaders() { req.setValue(v, forHTTPHeaderField: k) }
        let (data, _) = try await Net.data(req)
        let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        let out = json["output"] as? [String: Any] ?? [:]
        let state = (out["task_status"] as? String) ?? "UNKNOWN"
        let vurl = JSONValue.string(out, ["video_url"]) ?? JSONValue.string(out, ["results", "0", "url"])
        let msg = out["message"] as? String ?? json["message"] as? String
        return VideoStatus(state: state, videoUrl: vurl, message: msg, raw: json)
    }

    // MARK: - 文生图（同步 multimodal-generation）
    //
    // 接口：POST /services/aigc/multimodal-generation/generation
    // 请求体：{model, input:{messages:[{role:"user", content:[{text}]}]}, parameters:{size,n}}
    // 同步返回 output.choices[].message.content[].image（图片 URL，可能多张）
    // 注意：size 分隔符是「*」不是「x」，如 "1024*1024"

    struct ImageRequest {
        var prompt: String
        var model: String = FixedModel.imageDefault
        var size: String = "1024*1024"
        var n: Int = 1
        var promptExtend: Bool = true
        var watermark: Bool = false
    }

    /// 同步生成图片，返回图片 URL 数组（通常 1 张，wan 系列可能多张）
    func submitImage(_ r: ImageRequest) async throws -> [String] {
        let params: [String: Any] = [
            "size": r.size,
            "n": r.n,
            "watermark": r.watermark,
            "prompt_extend": r.promptExtend,
        ]
        let body: [String: Any] = [
            "model": r.model,
            "input": [
                "messages": [
                    ["role": "user", "content": [["text": r.prompt]]]
                ]
            ],
            "parameters": params,
        ]
        let json = try await postJSON(
            "\(DashScope.httpBase)/services/aigc/multimodal-generation/generation",
            body: body)

        // 解析 output.choices[*].message.content[*].image
        var urls: [String] = []
        if let out = json["output"] as? [String: Any],
           let choices = out["choices"] as? [[String: Any]] {
            for ch in choices {
                guard let msg = ch["message"] as? [String: Any],
                      let content = msg["content"] as? [[String: Any]] else { continue }
                for item in content {
                    if let u = item["image"] as? String, !u.isEmpty { urls.append(u) }
                }
            }
        }
        // 兼容某些模型返回 results 结构
        if urls.isEmpty {
            if let out = json["output"] as? [String: Any],
               let results = out["results"] as? [[String: Any]] {
                for item in results {
                    if let u = item["url"] as? String, !u.isEmpty { urls.append(u) }
                }
            }
        }
        if urls.isEmpty {
            throw APIError("生成失败：未返回图片地址 \(json["code"] ?? "") \(json["message"] ?? "")")
        }
        return urls
    }

    // MARK: - 下载远端资源到本地

    func download(_ remote: URL, to dest: URL) async throws {
        let (tmp, _) = try await Net.download(URLRequest(url: remote))
        if FileManager.default.fileExists(atPath: dest.path) {
            try? FileManager.default.removeItem(at: dest)
        }
        try FileManager.default.moveItem(at: tmp, to: dest)
    }

    // MARK: - 通用 POST JSON

    @discardableResult
    func postJSON(_ urlString: String, body: [String: Any], extra: [String: String] = [:]) async throws -> [String: Any] {
        guard let url = URL(string: urlString) else { throw APIError("URL 无效") }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.setValue("application/json", forHTTPHeaderField: "Accept")
        for (k, v) in try authHeaders(extra: extra) { req.setValue(v, forHTTPHeaderField: k) }
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, resp) = try await Net.data(req)
        let status = (resp as? HTTPURLResponse)?.statusCode ?? -1
        let json = (try? JSONSerialization.jsonObject(with: data)) as? [String: Any] ?? [:]
        if !(200..<300).contains(status) {
            let code = json["code"] as? String ?? ""
            let msg = json["message"] as? String ?? String(data: data, encoding: .utf8) ?? ""
            throw APIError("HTTP \(status) \(code) \(msg)".trimmingCharacters(in: .whitespaces))
        }
        return json
    }
}
